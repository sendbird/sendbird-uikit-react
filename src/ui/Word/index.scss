@import '../../styles/variables';

.sendbird-word {
  display: inline;
  margin-right: 4px;
  height: fit-content;
  white-space: nowrap;
  .sendbird-word__url,
  .sendbird-word__normal {
    display: inline-block;
    color: inherit;
  }
}
