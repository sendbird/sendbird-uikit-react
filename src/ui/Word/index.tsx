import React from 'react';
import SendBird from 'sendbird';
import './index.scss';

import { LabelTypography, LabelColors } from '../Label';
import LinkLabel from '../LinkLabel';
import uuidv4 from '../../utils/uuid';
import { convertWordToStringObj, StringObjType, StringObj } from '../../utils';
import MentionLabel from '../MentionLabel';

interface WordProps {
  word: string;
  message: SendBird.UserMessage;
  isByMe?: boolean;
  mentionTemplate?: string;
  renderString?: (stringObj: StringObj) => JSX.Element;
}

// Word and StringObj will include types: normal, mention, url
export default function Word(props: WordProps): JSX.Element {
  const {
    word,
    message,
    isByMe = false,
    mentionTemplate = '@',
    renderString = null,
  } = props;
  return (
    <span className="sendbird-word">
      {
        convertWordToStringObj(word, message?.mentionedUsers).map((stringObj) => {
          const { type, value } = stringObj;
          if (renderString && typeof renderString === 'function') {
            return renderString(stringObj);
          }
          if (type === StringObjType.mention) {
            return (
              <MentionLabel
                mentionTemplate={mentionTemplate}
                mentionedUserId={value}
                key={uuidv4()}
                isByMe={isByMe}
              />
            );
          } else if (type === StringObjType.url) {
            return (
              <LinkLabel
                className="sendbird-word__url"
                key={uuidv4()}
                src={word}
                type={LabelTypography.BODY_1}
                color={isByMe ? LabelColors.ONCONTENT_1 : LabelColors.ONBACKGROUND_1}
              >
                {value}
              </LinkLabel>
            );
          } else {
            return value
          }
        })
      }
    </span>
  );
}
