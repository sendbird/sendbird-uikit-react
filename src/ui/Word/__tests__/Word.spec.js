import React from 'react';
import { render } from '@testing-library/react';

import Word from "../index";

describe('ui/Word', () => {
  it('should do a snapshot test of the Word DOM', function() {
    const text = "example-text";
    const { asFragment } = render(
      <Word
        word="hello@{hoon}"
        message={{
          mentionedUsers: [{ userId: 'hoon', nickname: 'Hoon Baek' }]
        }}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
