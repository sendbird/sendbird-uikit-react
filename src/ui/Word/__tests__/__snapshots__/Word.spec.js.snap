// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Word should do a snapshot test of the Word DOM 1`] = `
<span
  className="sendbird-word"
>
  hello
  <div
    className="sendbird-context-menu"
    style={
      Object {
        "display": "inline",
      }
    }
  >
    <a
      className="
            sendbird-word__mention
            
          "
      onClick={[Function]}
    >
      <span
        className="sendbird-label sendbird-label--caption-1 sendbird-label--color-onbackground-1"
      >
        @Hoon Baek
      </span>
    </a>
  </div>
</span>
`;
